src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.cpp

```diff
@@ -65,6 +65,48 @@ ONNXOpShapeHelper<OP>::ONNXOpShapeHelper(OP *newOp, int numResults,
   };
 }
 
+template <>
+Value ONNXOpShapeHelper<Operation>::getOutput(int n) {
+  return op->getResult(n);
+}
+
+template <class OP>
+DimsExpr &ONNXOpShapeHelper<OP>::dimsForOutput(int n) {
+  Value output = getOutput(n);
+  // Try to combine the inferred shape and the output's shape if possbile.
+  if (isRankedShapedType(output.getType())) {
+    llvm::ArrayRef<int64_t> existingDims = getShape(output.getType());
+    // Do not handle the case of scalar tensor whose type can be tensor<f32>
+    // or tensor<1xf32>. Just use the inferredShape in this case.
+    if (existingDims.size() >= 1 && outputsDims[n].size() >= 1) {
+      if (existingDims.size() != outputsDims[n].size())
+        llvm::outs() << "Inferred rank (" << outputsDims[n].size()
+                     << ") vs. existing rank (" << existingDims.size() << ")\n";
+      assert((existingDims.size() == outputsDims[n].size()) &&
+             "Inferred shape and existing shape are inconsistent in the number "
+             "of elements");
+      for (unsigned i = 0; i < existingDims.size(); ++i) {
+        // existingDim is static, inferedDim is unknown: update the
+        // inferredDim.
+        if ((existingDims[i] != -1) && outputsDims[n][i].isQuestionmark())
```

Comment:
I would restructure 
```C++
if (existingDIms[i] != -1) {
  if (question mark) { 
    // we have question mark only during the shape inference
  } else if (!literal) {
     // NEW CASE: we have computation during the lowering... use the constant given by user
   } else if (literal) {
     // issue a warning if found a different value, keep user given value
  } 
}
```

Reply:
Done and added the new case so that the lowering code uses static dim instead. Hope it's what you mean. Thanks!